### ISSUE TITLE
(0.15.0-rc.1) "prepared writing fileset volume that already exists"

### ISSUE THREAD

**Issue by BertHartm** [m3db/m3#2193]
We're running one 0.15.0-rc.1 node in an otherwise 0.14.2 cluster. It was upgraded to see if it could make it bootstrap better as it was in a crashloop. It eventually was able to peer bootstrap.

We now see large amounts of log output of the form 
```
Mar  4 20:05:37 dbnode-816619873-4-823954132 m3dbnode: {"level":"error","ts":1583352337.5617883,"msg":"prepared writing fileset volume that already exists","violation":"invariant_violated","blockStart":1582891200,"fileSetType":"flush","volumeIndex":0,"snapshotStart":-6795364578.8713455,"namespace":"14d","shard":3550}
Mar  4 20:05:37 dbnode-816619873-4-823954132 m3dbnode: {"level":"error","ts":1583352337.564659,"msg":"prepared writing fileset volume that already exists","violation":"invariant_violated","blockStart":1582891200,"fileSetType":"flush","volumeIndex":0,"snapshotStart":-6795364578.8713455,"namespace":"14d","shard":3739}
Mar  4 20:05:37 dbnode-816619873-4-823954132 m3dbnode: {"level":"error","ts":1583352337.564659,"msg":"prepared writing fileset volume that already exists","violation":"invariant_violated","blockStart":1582891200,"fileSetType":"flush","volumeIndex":0,"snapshotStart":-6795364578.8713455,"namespace":"14d","shard":3739}
Mar  4 20:05:37 dbnode-816619873-4-823954132 m3dbnode: {"level":"error","ts":1583352337.5674076,"msg":"prepared writing fileset volume that already exists","violation":"invariant_violated","blockStart":1582891200,"fileSetType":"flush","volumeIndex":0,"snapshotStart":-6795364578.8713455,"namespace":"14d","shard":3934}
Mar  4 20:05:37 dbnode-816619873-4-823954132 m3dbnode: {"level":"error","ts":1583352337.5674076,"msg":"prepared writing fileset volume that already exists","violation":"invariant_violated","blockStart":1582891200,"fileSetType":"flush","volumeIndex":0,"snapshotStart":-6795364578.8713455,"namespace":"14d","shard":3934}
```
and 
```
...led to flush data: persist manager cannot prepare, fileset already exists\nshard 3071 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3073 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3074 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3078 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3081 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3083 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3084 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3085 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3092 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3093 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3094 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3095 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3097 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3098 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3099 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3101 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3102 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3104 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3105 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3106 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3108 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3109 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3112 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3113 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3114 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3116 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3120 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3122 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3125 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3129 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3130 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3131 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3133 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3134 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3135 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3136 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3137 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3139 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3141 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3142 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3143 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3144 failed to flush data: persist manager cannot prepare, fileset already exists\nshard 3150 failed to flush data: persist manager cannot prepare,...
```
(not actually sure where that line begins or ends).

the node otherwise seems healthy.

## General Issues

General issues are any non-performance related issues (data integrity, ease of use, error messages, configuration, documentation, etc).

Please provide the following information along with a description of the issue that you're experiencing:

1. What service is experiencing the issue? (M3Coordinator, M3DB, M3Aggregator, etc)
M3DB
2. What is the configuration of the service? Please include any YAML files, as well as namespace / placement configuration (with any sensitive information anonymized if necessary).
6 nodes in an RF3 placement, 2 namespaces set for 14d retention, 6 hour blocks.
3. How are you using the service? For example, are you performing read/writes to the service via Prometheus, or are you using a custom script?
Primarily prometheus writes.
4. Is there a reliable way to reproduce the behavior? If so, please provide detailed instructions.
Not sure, going to see if it persists after a reboot.

**Comment by BertHartm** [comment]
After a reboot the issue has gone away. There was a warning about corrupt snapshot in the logs:
```
{"level":"warn","ts":1583371489.535426,"msg":"corrupt snapshot file during cleanup, marking files for deletion","comment":"partial/corrupt files are expected as result of a restart (this is ok)","error":"error reading snapshot info file: open /data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-checkpoint.db: no such file or directory","files":["/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-bloomfilter.db","/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-data.db","/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-digest.db","/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-index.db","/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-info.db","/data/m3db/snapshots/14d/1416/fileset-1583344800000000000-28-summaries.db"]}
```
but as it's just the one shard, I don't know if it's related or not.

**Comment by bismarck** [comment]
I just saw this same issue on a cluster running 0.15.0-rc.3.  I removed a node from a placement and then added a new node.  Started to see log output noted above on the new node and 3 other nodes in the cluster (probably the source of the shards).  I wonder if this is related to #2294?

**Comment by gibbscullen** [comment]
Fixed with #2338
